src/Core/Utilities/CustomRedisProcessingStrategy.cs

```diff
@@ -0,0 +1,128 @@
+/// the default implementation would throw an exception and exit the request pipeline for all requests. 
+/// </remarks>
+public class CustomRedisProcessingStrategy : ProcessingStrategy
+{
+    private readonly IConnectionMultiplexer _connectionMultiplexer;
+    private readonly IRateLimitConfiguration _config;
+    private readonly ILogger<CustomRedisProcessingStrategy> _logger;
+    private readonly IMemoryCache _memoryCache;
+    private readonly GlobalSettings.DistributedIpRateLimitingSettings _distributedSettings;
+
+    private const string _redisTimeoutCacheKey = "IpRateLimitRedisTimeout";
+
+    public CustomRedisProcessingStrategy(
+        IConnectionMultiplexer connectionMultiplexer,
+        IRateLimitConfiguration config,
+        ILogger<CustomRedisProcessingStrategy> logger,
+        IMemoryCache memoryCache,
+        GlobalSettings globalSettings)
+        : base(config)
+    {
+        _connectionMultiplexer = connectionMultiplexer ??
+                                 throw new ArgumentException(
+                                     "IConnectionMultiplexer was null. Ensure StackExchange.Redis was successfully registered");
+        _config = config;
+        _logger = logger;
+        _memoryCache = memoryCache;
+        _distributedSettings = globalSettings.DistributedIpRateLimiting;
+    }
+
+    private static readonly LuaScript _atomicIncrement = LuaScript.Prepare(
+        "local count = redis.call(\"INCRBYFLOAT\", @key, tonumber(@delta)) local ttl = redis.call(\"TTL\", @key) if ttl == -1 then redis.call(\"EXPIRE\", @key, @timeout) end return count");
+
+    public override async Task<RateLimitCounter> ProcessRequestAsync(ClientRequestIdentity requestIdentity,
+        RateLimitRule rule, ICounterKeyBuilder counterKeyBuilder, RateLimitOptions rateLimitOptions,
+        CancellationToken cancellationToken = default)
+    {
+        // If Redis is down entirely, skip rate limiting
+        if (!_connectionMultiplexer.IsConnected)
+        {
+            _logger.LogDebug("Redis connection is down, skipping IP rate limiting");
+            return SkipRateLimitResult();
+        }
+
+        // Check if any Redis timeouts have occured recently
+        if (_memoryCache.TryGetValue<TimeoutCounter>(_redisTimeoutCacheKey, out var timeoutCounter))
+        {
+            // We've exceeded threshold, backoff Redis and skip rate limiting for now
+            if (timeoutCounter.Count >= _distributedSettings.MaxRedisTimeoutsThreshold)
+            {
+                _logger.LogDebug(
+                    "Redis timeout threshold has been exceeded, backing off and skipping IP rate limiting");
+                return SkipRateLimitResult();
+            }
+        }
+
+        var counterId = BuildCounterKey(requestIdentity, rule, counterKeyBuilder, rateLimitOptions);
+
+        try
+        {
+            return await IncrementAsync(counterId, rule.PeriodTimespan ?? rule.Period.ToTimeSpan(), _config.RateIncrementer);
```

Comment:
can we just call `base.ProcessRequestAsync` here instead?

Reply:
Yep, that does make a lot more sense